### PR TITLE
[OpenMP][MLIR][OMPIRBuilder] Add a small optional constant alloca raise function pass to finalize, utilised in convertTarget

### PR DESCRIPTION
**Pull request by agozillon**
This patch seeks to add a mechanism to raise constant (not ConstantExpr or runtime/dynamic) sized allocations into the entry block for select functions that have been inserted into a list for processing. This processing occurs during the finalize call, after OutlinedInfo regions have completed. This currently has only been utilised for createOutlinedFunction, which is triggered for TargetOp generation in the OpenMP MLIR dialect lowering to LLVM-IR.

This currently is required for Target kernels generated by createOutlinedFunction to avoid subsequent optimization passes doing some unintentional malformed optimizations for AMD kernels (unsure if it occurs for other vendors). If the allocas are generated inside of the kernel and are not in the entry block and are subsequently passed to a function this can lead to required instructions being erased or manipulated in a way that causes the kernel to run into a HSA access error.

This fix is related to a series of problems found in: https://github.com/llvm/llvm-project/issues/74603 

This problem primarily presents itself for Flang's HLFIR AssignOp currently, when utilised with a scalar temporary constant on the RHS and a descriptor type on the LHS. It will generate a call to a runtime function, wrap the RHS temporary in a newly allocated descriptor (an llvm struct), and pass both the LHS and RHS descriptor into the runtime function call. This will currently be
embedded into the middle of the target region in the user entry block, which means the allocas are also embedded in the middle, which seems to pose
issues when later passes are executed. This issue may present itself in other HLFIR operations or unrelated operations that generate allocas as a by product, but for the moment, this one test case is the only scenario I've found this problem.

Perhaps this is not the appropriate fix, I am very open to other suggestions, I've tried a few others (at varying levels of the flang/mlir compiler flow), but this one is the smallest and least intrusive change set. The other two, that come to mind (but I've not fully looked into, the former I tried a little with blocks but it had a few issues I'd need to think through):

-  Having a proper alloca only block (or region) generated for TargetOps that we could merge into the entry block that's generated by convertTarget's createOutlinedFunction.
- Or diverging a little from Clang's current target generation and using the CodeExtractor to generate the user code as an outlined function region invoked from the kernel we make, with our kernel arguments passed into it. Similar to the current parallel generation. I am not sure how well this would intermingle with the existing parallel generation though that's layered in.

Both of these methods seem like quite a divergence from the current status quo, which I am not entirely sure is merited for the small test this change aims to fix.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
Could you share the HLFIR IR that needs this change?

**Comment by agozillon** [comment]
> Could you share the HLFIR IR that needs this change?

Sure, here you go: 

```
module attributes {dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<i1, dense<8> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr, dense<64> : vector<4xi64>>, #dlti.dl_entry<f16, dense<16> : vector<2xi64>>, #dlti.dl_entry<i32, dense<32> : vector<2xi64>>, #dlti.dl_entry<i16, dense<16> : vector<2xi64>>, #dlti.dl_entry<i8, dense<8> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr<272>, dense<64> : vector<4xi64>>, #dlti.dl_entry<!llvm.ptr<271>, dense<32> : vector<4xi64>>, #dlti.dl_entry<!llvm.ptr<270>, dense<32> : vector<4xi64>>, #dlti.dl_entry<f128, dense<128> : vector<2xi64>>, #dlti.dl_entry<f64, dense<64> : vector<2xi64>>, #dlti.dl_entry<f80, dense<128> : vector<2xi64>>, #dlti.dl_entry<i128, dense<128> : vector<2xi64>>, #dlti.dl_entry<i64, dense<64> : vector<2xi64>>, #dlti.dl_entry<"dlti.stack_alignment", 128 : i64>, #dlti.dl_entry<"dlti.endianness", "little">>, fir.defaultkind = "a1c4d8i4l4r4", fir.kindmap = "", llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128", llvm.target_triple = "x86_64-unknown-linux-gnu", omp.is_gpu = false, omp.is_target_device = false, omp.requires = #omp<clause_requires none>, omp.version = #omp.version<version = 11>} {
  func.func @_QQmain() attributes {fir.bindc_name = "main"} {
    %0 = fir.alloca !fir.box<!fir.heap<i32>> {bindc_name = "test", uniq_name = "_QFEtest"}
    %1 = fir.zero_bits !fir.heap<i32>
    %2 = fir.embox %1 : (!fir.heap<i32>) -> !fir.box<!fir.heap<i32>>
    fir.store %2 to %0 : !fir.ref<!fir.box<!fir.heap<i32>>>
    %3:2 = hlfir.declare %0 {fortran_attrs = #fir.var_attrs<allocatable>, uniq_name = "_QFEtest"} : (!fir.ref<!fir.box<!fir.heap<i32>>>) -> (!fir.ref<!fir.box<!fir.heap<i32>>>, !fir.ref<!fir.box<!fir.heap<i32>>>)
    %4 = fir.allocmem i32 {fir.must_be_heap = true, uniq_name = "_QFEtest.alloc"}
    %5 = fir.embox %4 : (!fir.heap<i32>) -> !fir.box<!fir.heap<i32>>
    fir.store %5 to %3#1 : !fir.ref<!fir.box<!fir.heap<i32>>>
    %6 = fir.load %3#1 : !fir.ref<!fir.box<!fir.heap<i32>>>
    %7 = fir.box_offset %3#1 base_addr : (!fir.ref<!fir.box<!fir.heap<i32>>>) -> !fir.llvm_ptr<!fir.ref<i32>>
    %8 = fir.box_addr %6 : (!fir.box<!fir.heap<i32>>) -> !fir.heap<i32>
    %9 = omp.map_info var_ptr(%8 : !fir.heap<i32>, i32) var_ptr_ptr(%7 : !fir.llvm_ptr<!fir.ref<i32>>) map_clauses(tofrom) capture(ByRef) -> !fir.heap<i32> {name = "test"}
    %10 = omp.map_info var_ptr(%3#1 : !fir.ref<!fir.box<!fir.heap<i32>>>, !fir.box<!fir.heap<i32>>) map_clauses(tofrom) capture(ByRef) members(%9 : !fir.heap<i32>) -> !fir.ref<!fir.box<!fir.heap<i32>>> {name = "test"}
    omp.target map_entries(%10 -> %arg0 : !fir.ref<!fir.box<!fir.heap<i32>>>) {
    ^bb0(%arg0: !fir.ref<!fir.box<!fir.heap<i32>>>):
      %20:2 = hlfir.declare %arg0 {fortran_attrs = #fir.var_attrs<allocatable>, uniq_name = "_QFEtest"} : (!fir.ref<!fir.box<!fir.heap<i32>>>) -> (!fir.ref<!fir.box<!fir.heap<i32>>>, !fir.ref<!fir.box<!fir.heap<i32>>>)
      %c50_i32 = arith.constant 50 : i32
      hlfir.assign %c50_i32 to %20#0 realloc : i32, !fir.ref<!fir.box<!fir.heap<i32>>>
      omp.terminator
    }
    %c6_i32 = arith.constant 6 : i32
    %11 = fir.address_of(@_QQclXf0a9a187466a9dd5699bdae5c0b697a3) : !fir.ref<!fir.char<1,81>>
    %12 = fir.convert %11 : (!fir.ref<!fir.char<1,81>>) -> !fir.ref<i8>
    %c9_i32 = arith.constant 9 : i32
    %13 = fir.call @_FortranAioBeginExternalListOutput(%c6_i32, %12, %c9_i32) fastmath<contract> : (i32, !fir.ref<i8>, i32) -> !fir.ref<i8>
    %14 = fir.load %3#0 : !fir.ref<!fir.box<!fir.heap<i32>>>
    %15 = fir.box_addr %14 : (!fir.box<!fir.heap<i32>>) -> !fir.heap<i32>
    %16 = fir.load %15 : !fir.heap<i32>
    %c2_i32 = arith.constant 2 : i32
    %17 = arith.addi %16, %c2_i32 : i32
    %18 = fir.call @_FortranAioOutputInteger32(%13, %17) fastmath<contract> : (!fir.ref<i8>, i32) -> i1
    %19 = fir.call @_FortranAioEndIoStatement(%13) fastmath<contract> : (!fir.ref<i8>) -> i32
    return
  }
  fir.global linkonce @_QQclXf0a9a187466a9dd5699bdae5c0b697a3 constant : !fir.char<1,81> {
    %0 = fir.string_lit "/home/agozillo/git/flang-dev/work-dir/declare-target-map/single-value-alloca.f90\00"(81) : !fir.char<1,81>
    fir.has_value %0 : !fir.char<1,81>
  }
  func.func private @_FortranAioBeginExternalListOutput(i32, !fir.ref<i8>, i32) -> !fir.ref<i8> attributes {fir.io, fir.runtime}
  func.func private @_FortranAioOutputInteger32(!fir.ref<i8>, i32) -> i1 attributes {fir.io, fir.runtime}
  func.func private @_FortranAioEndIoStatement(!fir.ref<i8>) -> i32 attributes {fir.io, fir.runtime}
  fir.global @_QQEnvironmentDefaults constant : !fir.ref<tuple<i32, !fir.ref<!fir.array<0xtuple<!fir.ref<i8>, !fir.ref<i8>>>>>> {
    %0 = fir.zero_bits !fir.ref<tuple<i32, !fir.ref<!fir.array<0xtuple<!fir.ref<i8>, !fir.ref<i8>>>>>>
    fir.has_value %0 : !fir.ref<tuple<i32, !fir.ref<!fir.array<0xtuple<!fir.ref<i8>, !fir.ref<i8>>>>>>
  }
}
```

The relevant location is inside of the TargetOp, the rest is largely irrelevant for device but included for completeness. The HLFIR AssignOp will generate an allocation as part of it's lowering from my understanding (in this particular use case at least, there's different lowering's for it). Although, I wouldn't say the problem lies with the HLFIR operation necessarily, more just a side affect of how we lower to a target region for LLVM-IR being shown I think (maybe there's a way to make it more AMDGPU runtime friendly, or a different runtime function for AMDGPU, but I'm doubtful that's the ideal solution). Even if we make sure to raise all AllocaOp's (when we lower to the LLVM Dialect this is done currently, but TargetOp doesn't generate a new Blocks, so there is no isolated Entry Block as such) to the top this would still persist, as we inevitably embed the user code into a seperate block in-between some kernel entry code for the arguments that will branch off to a fail condition or the user code block containing the allocations, and then a later pass (not sure which yet unfortunately) will try to do some magic and end up breaking the generated executable.

**Comment by agozillon** [comment]
Probably worth noting this issue doesn't exist for the current variations of target parallel from what I've found, likely because the CodeExtractor magics the user code into a seperate function with an appropriate entry block (my knowledge on the lowering of parallel is still a WIP, so please take it with a grain of salt and anyone more familiar please feel free to override my statement).

**Comment by agozillon** [comment]
Small ping for some reviewer attention next week if any time can be spared

**Comment by agozillon** [comment]
Another small ping for reviewer attention if at all possible please.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (TIFitis) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:648]: Nit: Remove llvm::

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (TIFitis) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:642]: The wording of this comment is a little off. Can you please rephrase it.

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h
Comment (TIFitis) [inline llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:1513]: Does this really need to be a separate function?

**Review by TIFitis — APPROVED** [review]
Overall seems fine to me. However I'm not familiar about when and why this is required, so please wait for other reviewers as well.

Also, should we have a test for this in the OMPIRBuilderTest?

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h
Comment (agozillon) [reply llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:1513]: I don't think so if we don't wish it to be, it was primarily to mimic the current style we use for addOutlineInfo if I recall correctly. Happy to change it!

**Comment by agozillon** [comment]
Thank you very much for the review @TIFitis I'll update this at the beginning of next week. And I'll hold off until we get a second reviewer that can sign off.

**Comment by agozillon** [comment]
Updated PR based on recent comments [(commit)](https://github.com/llvm/llvm-project/pull/78818/commits/f568ba2ab00a42bb5fdc7599ce60f03955605f4e) rebased on recent upstream changes to make sure it remains functional.

Changes made were: 

- Fix badly written comment
- Remove unnecessary function
- remove unrequired llvm:: prefix

Still looking for another reviewer for this PR if possible or further discussion.

**Comment by agozillon** [comment]
It slipped my mind that @TIFitis asked for an OpenMPIRBuilderTest, so I have now added one testing the alloca raising that this PR seeks to add in the latest commit!

**Comment by kiranchandramohan** [comment]
> The relevant location is inside of the TargetOp, the rest is largely irrelevant for device but included for completeness. The HLFIR AssignOp will generate an allocation as part of it's lowering from my understanding (in this particular use case at least, there's different lowering's for it). Although, I wouldn't say the problem lies with the HLFIR operation necessarily, more just a side affect of how we lower to a target region for LLVM-IR being shown I think (maybe there's a way to make it more AMDGPU runtime friendly, or a different runtime function for AMDGPU, but I'm doubtful that's the ideal solution). Even if we make sure to raise all AllocaOp's (when we lower to the LLVM Dialect this is done currently, but TargetOp doesn't generate a new Blocks, so there is no isolated Entry Block as such) to the top this would still persist, as we inevitably embed the user code into a seperate block in-between some kernel entry code for the arguments that will branch off to a fail condition or the user code block containing the allocations, and then a later pass (not sure which yet unfortunately) will try to do some magic and end up breaking the generated executable.

Is your observation that the OpenMPIRBuilder creates a separate `Alloca` block and this `Alloca` block will become the entry block in LLVM IR. So even if we move all the allocas to the entry block in FIR/HLFIR/LLVM dialect MLIR, it still won't be the entry block in LLVM IR? 

Is it possible to merge the `Alloca` block to the succeeding block (which is hopefully the entry block at the MLIR dialects)? And if we have lifted in MLIR all to the entry block this will all work out fine?

**Comment by agozillon** [comment]
> > The relevant location is inside of the TargetOp, the rest is largely irrelevant for device but included for completeness. The HLFIR AssignOp will generate an allocation as part of it's lowering from my understanding (in this particular use case at least, there's different lowering's for it). Although, I wouldn't say the problem lies with the HLFIR operation necessarily, more just a side affect of how we lower to a target region for LLVM-IR being shown I think (maybe there's a way to make it more AMDGPU runtime friendly, or a different runtime function for AMDGPU, but I'm doubtful that's the ideal solution). Even if we make sure to raise all AllocaOp's (when we lower to the LLVM Dialect this is done currently, but TargetOp doesn't generate a new Blocks, so there is no isolated Entry Block as such) to the top this would still persist, as we inevitably embed the user code into a seperate block in-between some kernel entry code for the arguments that will branch off to a fail condition or the user code block containing the allocations, and then a later pass (not sure which yet unfortunately) will try to do some magic and end up breaking the generated executable.
> 
> Is your observation that the OpenMPIRBuilder creates a separate `Alloca` block and this `Alloca` block will become the entry block in LLVM IR. So even if we move all the allocas to the entry block in FIR/HLFIR/LLVM dialect MLIR, it still won't be the entry block in LLVM IR?

That's correct unfortunately with our current lowering method (which I believe is just how we traditionally generate the kernels from what I can gather from code comments) we generate an initial entry point that does some load in/setup of kernel input and then performs a libomptarget runtime invocation for initialization (__kmpc_target_init) before branching into a fail condition or the user entry code, which is effectively the lowered MLIR code in the region. So any allocas in the region will still be emitted outside of the alloca block currently. 

> Is it possible to merge the `Alloca` block to the succeeding block (which is hopefully the entry block at the MLIR dialects)? And if we have lifted in MLIR all to the entry block this will all work out fine?

I thought about doing this, but there's a few issues (some may be simpler to overcome than I think or I maybe my thinking is just bad) I thought of that might get in the way of it:

1) I don't think we can just merge the MLIR generated block into the front of the newly generated OpenMPIRBuilder target block as we could end up with out of synch operations, if anything in the MLIR generated block depends on any of the input arguments. 
2) The inverse of merging at the bottom of the block becomes a bit of an issue due to the kmpc_target_init call into a branch, as we can't meld it directly behind the branch. 
3) We could maybe splice it into the OpenMPIRBuilder generated entry block, above the call to kmpc_target_init and branch below the argument entry code. However, we actually do not really have the concept of an strict alloca block when it comes to TargetOp (or any of the OpenMP region constructs from what I can tell), we simply prepend allocas to the top of the target region, and this primarily seems to be enforced in the FIR -> LLVM dialect lowering, the HLFIR/FIR CodeGen seems less strict on this. Enforcing a specific block at the FIR -> LLVM dialect lowering level didn't seem particularly hard when I tried it (but I didn't test it extensively). However, as I recall after doing this I found that it's not so easy to reliably find this alloca block when we're trying to merge them, as we can' t guarantee the first generated LLVM-IR block is the alloca block. So we'd need some way to identify the block for merging. 

We also need to be a tad careful when doing this as we have the nested parallel operations to consider that also do their own alloca movements and extraction, it's why I opted to defer the raising to the finalize method and restrict it solely to the Target kernel entry function. 

It's been a little bit since I looked into this in detail now, so hopefully I am recollecting the above well enough! I am happy to look into it again to recollect more details if that's something we'd be interested in doing.

**Comment by agozillon** [comment]
Just a small ping to check if you're happy with the fix as is @kiranchandramohan or if you'd prefer me to dig into the block merge/splice method to see if it's feasible?

File: mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir
Comment (kiranchandramohan) [inline mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir:29]: If there is a loop around this block, would this still move the alloca to the entry block? If so, is that still semantically equivalent?

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (kiranchandramohan) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5095]: Assuming this code is going to run for all device targets. I guess you said that this is something specific for AMD GPU backends. Do you know whether such transformations apply to other targets as well?

**Comment by kiranchandramohan** [comment]
> Just a small ping to check if you're happy with the fix as is @kiranchandramohan or if you'd prefer me to dig into the block merge/splice method to see if it's feasible?

No need to pursue this.

I have couple of questions inline. Also, would this be better as an LLVM pass?

File: mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir
Comment (agozillon) [reply mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir:29]: Thank you for the great question @kiranchandramohan and I encourage anyone else that has any thoughts on possible scenarios that this might break to also bring things up where there may be concerns. My LLVM knowledge is still very much a WIP.

In this case it would move the alloca to the entry block whether or not it's in a loop related series of blocks. From an MLIR perspective currently I don't think it's possible to have a series of loop blocks around the initial basic block of `TargetOp `(but I may be wrong on this and would love to know a scenario where this is possible to test), although, it is possible to have blocks chained from it relating to loops as normal. I believe even in the case of `omp.parallel`/`omp.wsloop` it's internal to the initial basic block, and this code actively avoids manipulating this kind of nested region currently, the entirety of the parallel do should be outlined into a seperate function by the time we try to raise allocations, the raising aims to only happen inside of the kernel entry function (`TargetOp`) currently, the parallel do has it's own allocation raising/sinking that occurs via the `CodeExtractor `that we need to avoid tampering with (as it will break it).

From a Fortran perspective (as someone who is still very bad at writing Fortran) I am not sure a user can inject an alloca directly into the `TargetOp`, from my understanding you can only declare a variable at the top of a function or in a module. So you can't quite do what you can in C++ where as an example you declare an integer directly in the OpenMP target region which will become an alloca. 

However, as we do still need to consider this being used in C/C++ where it is possible, if we look at the below C/C++ examples where I've tried to come up with some breaking scenarios (perhaps my logic isn't quite sound and there is better examples that would cause breakage, in which case please do bring them up!):

Version 1

```
int test = 0;
#pragma omp target map(tofrom : test)
{
  for (int i = 0; i < 10; i++) {
    int j = 10;
    j += 10;
    test += j;
  }
}
```

Version 2

```
int test = 0;
#pragma omp target map(tofrom : test)
{
  int j = 10;
  for (int i = 0; i < 10; i++) {
    j += 10;
    test += j;
  }
}
```

In both cases, when we emit the device IR the alloca for `j` is moved to the entry block alongside `i`, and the correct results are achieved, so I believe it is still semantically equivalent. Without the initialization of `j` with 10 we get the same UB in both cases as well.

The main case where I can see it being an issue (and there may be others, which I would love to hear if someone can think of them so I can investigate) is if we have an allocation that is size dependent on some loop dependent variable, making it a dynamic allocation, however, this PR doesn't tackle this case (and doesn't handle constant expressions either, it needs them folded into a constant value currently), it'd be a lot more complex to deal with and Flang hasn't presented a case where we need to handle this scenario yet (hopefully never). If we get to the point of requiring this it'd require a significantly more complex solution I think.

However, disclaimer again, I am not the most LLVM savvy person yet, so anyone that may be reading, please do feel free to mention cases that it may be an issue that you can think of or if any of the above may be incorrect.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (agozillon) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5095]: I unfortunately can only speak to AMD GPU backends as it's all I've tested the fortran offload examples I work with on and don't know enough about the other backends (e.g. NVIDIA) to really comment on them unfortunately. 

I can see if I can restrict it to the AMD target triple if that's something you'd like me to do to be cautious however? not sure how complicated it'd be to do so here, but I imagine it wouldn't be. 

I believe someone did manage to get some of the offloading working for NVIDIA GPU's a while ago if I recall correctly, but I am not sure how easy that was for them, so I could in theory test that if we'd like to as well.

**Comment by agozillon** [comment]
> > Just a small ping to check if you're happy with the fix as is @kiranchandramohan or if you'd prefer me to dig into the block merge/splice method to see if it's feasible?
> 
> No need to pursue this.
> 

Sounds good, if it ever becomes necessary to pursue the other option I am more than happy to do so.

> I have couple of questions inline. Also, would this be better as an LLVM pass?

I would like to avoid it being an LLVM pass for now if that's at all possible, it's a rather small/trivial piece of functionality that only gets performed on the Target/kernel entry function at the moment, I'd rather other people not start (or at least be able to) running it outside of the intended use-case for the moment. I think the easier to revert it is the better it is for the time being if we have to look at other options. However, if it expands to more complex use-cases (e.g. dynamic allocations or non-foldable constants, although, if we have to cover these cases I'd lean towards the entry block splicing or an alternative) then certainly it should be moved to an LLVM pass.

File: mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir
Comment (kiranchandramohan) [reply mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir:29]: > From a Fortran perspective (as someone who is still very bad at writing Fortran) I am not sure a user can inject an alloca directly into the `TargetOp`, from my understanding you can only declare a variable at the top of a function or in a module. So you can't quite do what you can in C++ where as an example you declare an integer directly in the OpenMP target region which will become an alloca.

There is a block construct in Fortran 2008 that can be used for this.

Comment (kiranchandramohan) [reply mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir:29]: Just pinging @jdoerfert if he has any opinion here.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (kiranchandramohan) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5095]: We do not have target specific code here at the moment. So no need to check this here.

**Review by kiranchandramohan — APPROVED** [review]
I think constant allocas are fine to be hoisted. But as mentioned elsewhere there could be other cases where allocas might cause problems and might require a generic solution.

If there are no concerns expressed in a day or two you can consider submitting this patch and generalize this later on.

File: mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir
Comment (agozillon) [reply mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir:29]: > > From a Fortran perspective (as someone who is still very bad at writing Fortran) I am not sure a user can inject an alloca directly into the `TargetOp`, from my understanding you can only declare a variable at the top of a function or in a module. So you can't quite do what you can in C++ where as an example you declare an integer directly in the OpenMP target region which will become an alloca.
> 
> There is a block construct in Fortran 2008 that can be used for this.
> 

Thank you Kiran! I can give this a shot then (provided it currently works in a target region, not tested it myself directly, so I am a little interested in that). Sorry that my Fortran knowledge is rather lacking in areas and thank you for introducing me to new stuff! 

> Also we have passes that can convert to Allocs to Allocas.
> 
> https://github.com/llvm/llvm-project/blob/2d0137dd64017c34101f8a763fd8958c20acd6bb/flang/lib/Optimizer/Transforms/StackArrays.cpp#L187

I'll have a look into the pass, it could be quite useful in the future. I did notice it when I was looking around for other possible solutions and it caught my eye, but I wasn't quite sure if we needed to look into an extensive solution just yet. While I don't think the pass would solve the issue (perhaps I am incorrect with my thinking however), it would be a good basis to create a pass that might. 

However, I think if we wanted to do this pass at an MLIR level we'd have to enforce an entry block for Target regions, and I believe it may be unnecessary to be a pass at that point, during the lowering to LLVM Dialect I think we try to move all allocas into the Target (or any) entry block, and if we make sure we have or generate an explicit entry block at this point, I think all allocas should hopefully end up naturally in the entry block where it's possible at least.

Comment (agozillon) [reply mlir/test/Target/LLVMIR/omptarget-constant-alloca-raise.mlir:29]: I made the following small little test case using blocks, it makes some attempt to generate the same behavior as the C++ test case above, with some extra on top:

```
PROGRAM main
    integer, allocatable :: a
    allocate(a)
    a = 5
 
!$omp target map(tofrom:a)
   sub_call : BLOCK
      INTEGER :: b
      b = a + 2
      CALL Sub(b)

      do i = 1, 10
        BLOCK
           integer :: j 
           j = 10
           j = j + 10
           a = a + j 
        END BLOCK
        a = a + b
     end do
    END BLOCK sub_call
!$omp end target

print *, a
END PROGRAM main
  
SUBROUTINE Sub(B)
  INTEGER :: b
  b = b * b
END SUBROUTINE Sub
```

It yields the same answer for both target and regular Fortran host (as does moving j outside and into the first block yield identical results), what is interesting is that forcing undefined behavior via removing the initial assignment to `j`, the regular Fortran host code will print a garbage value, whereas the target will return an actual consistent value, effectively not emitting the addition of `j` in the loop to `a`. This happens with or without the alloca raise pass here, so seems a seperate inconsistency.

I did, however, have to use the deprecated fir flow via `-flang-deprecated-no-hlfir` (which still uses the alloca raise, so it would still produce the relevant inconsistencies), as this little test has unfortunately shown another very similar compiler crash for allocatables to the one that this PR fixes (crashes with or without this PR sadly). So that will be fun to look into, I am unsure if it's a related or entirely new issue at this point though (will look into it more in the near future).

**Comment by agozillon** [comment]
Just a little forewarning, I will land this PR tomorrow, so if anyone has any concerns now would be the time to bring them forward! However, we can always revert the PR if there is afterwards (or any unexpected issues), but bringing them up just now would be ideal :-)

**Pull request merged**
